Automattic/jetpack
# Connection: extracting the Jetpack's authorization webhook

Fixes #15932.

The `authorize` webhook is used as a fallback action for Calypso authorize.
This PR is moving this webhook from Jetpack to `jetpack-connection`, making it available for plugins using the package without the Jetpack plugin.

#### Changes proposed in this Pull Request:
This commit introduces following changes:
- The `authorize` webhook is moved from Jetpack plugin into the Connection package.
- New class `Webhooks` responsible for handling the `authorize` and other requests we will move from Jetpack or create from scratch.
- Refactoring of the `Jetpack_Client_Server` class, deprecating some of its methods.
- Adjusting the existing `authorize` action handler to make it use the new `Webhooks` class, and deprecating it in favor of the new webhook.
- Unit tests for the new `Webhooks` class.

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
1. Switch Jetpack Beta to the branch, then disable Jetpack.
2. Install Client Example, run its `composer install` if needed. Delete its `vendor` directory, and symlink the one from `jetpack-dev`:
```bash
cd wp-content/plugins/client-example
composer install
rm -rf vendor
ln -s ../jetpack-dev/vendor
```
3. Go to "Client Example -> Simple UI with Calypso", click "Set Up Connection", but **do not** click "Approve".
4. Open `wp-config.php`, add `die();` into the beginning of the file, and click "Approve". You should see the `We had trouble connecting.` error message, and the "Try again" link.
5. Remove the `die();` from the file, and click on the "Try again" link. You should get redirected to the WP-Admin Dashboard. Go to "Client Example" and confirm that you're fully connected (both "Blog token" and "user token sections indicate that you're connected).
6. Activate the Jetpack plugin, disconnect it.
7. Go to "Settings -> Jetpack Constants" and enable the `JETPACK_SHOULD_NOT_USE_CONNECTION_IFRAME` constant.
8. Go to Jetpack Dashboard and connect Jetpack. After going through the Calypso connection flow, you should get redirected to Jetpack Dashboard with Jetpack fully connected.
9. Enable the `WP_DEBUG` constant in `wp-config.php`.
10. Load the URL `wp-admin/admin.php?page=jetpack&action=authorize`. You should see the `webhook is deprecated` error notice.
11. Disable the `WP_DEBUG` constant and reload that URL. You should get redirected to Jetpack Dashboard with the message "Your Jetpack is already connected."

#### Proposed changelog entry for your changes:
_Not sure if needed, but anyway:_
- Resolved an issue with connecting plugins to WordPress.com.
